File: cs/days/day21.cs
namespace Shunty.AoC.Days;

// https://adventofcode.com/2024/day/21 - Keypad Conundrum

public class Day21 : AocDaySolver
{
    public int DayNumber => 21;
    public string Title => "Keypad Conundrum";

/*

+---+---+---+
| 7 | 8 | 9 |
+---+---+---+
| 4 | 5 | 6 |
+---+---+---+
| 1 | 2 | 3 |
+---+---+---+
    | 0 | A |
    +---+---+

    +---+---+
    | ^ | A |
+---+---+---+
| < | v | > |
+---+---+---+

 */

    // Hard code the best paths for the direction pad 'cos there aren't that many. We'll add all paths for the
    // number pad in AddNumpadPaths() call. We'll ignore zig-zag paths such as "<^<".
    // Must take care not to include paths that cross the empty space.
    private static Dictionary<string, List<string>> AllPadPaths = new() {
        { "AA", ["A"] },            { "A^", ["<A"] },         { "A<", ["v<<A"] },  { "Av", ["<vA", "v<A"] }, { "A>", ["vA"] },
        { "^A", [">A"] },           { "^^", ["A"] },          { "^<", ["v<A"] },   { "^v", ["vA"] },         { "^>", ["v>A", ">vA"] },
        { "<A", [">>^A"] },         { "<^", [">^A"] },        { "<<", ["A"] },     { "<v", [">A"] },         { "<>", [">>A"] },
        { "vA", ["^>A", ">^A"] },   { "v^", ["^A"] },         { "v<", ["<A"] },    { "vv", ["A"] },          { "v>", [">A"] },
        { ">A", ["^A"] },           { ">^", ["<^A", "^<A"] }, { "><", ["<<A"] },   { ">v", ["<A"] },         { ">>", ["A"] },
    };

    public async Task Solve()
    {
        var input = (await File.ReadAllLinesAsync(AocUtils.FindInputFile(DayNumber))).ToList();
        //var input = (await Task.FromResult(TestInput.Trim())).Split('\n').Select(s => s.Trim()).ToList();

        AddNumpadPaths();

        long part1 = 0, part2 = 0;
        foreach (var line in input)
        {
            var p1 = PathLength(line, 3);
            var p2 = PathLength(line, 26);
            //AnsiConsole.WriteLine($"{line} => {p1}; {p2}");
            part1 += p1 * int.Parse(line[..^1]);
            part2 += p2 * int.Parse(line[..^1]);
        }
        this.ShowDayResult(1, part1);
        this.ShowDayResult(2, part2);
    }

    private static Dictionary<(string, int), long> Cache = [];
    private static long PathLength(string path, int level)
    {
        if (level == 0)
            return path.Length;
        if (Cache.TryGetValue((path, level), out long result))
            return result;

        result = 0;
        for (var i = 0; i < path.Length; i++)
        {
            // Split the path into 2 character chunks that we can look up in the paths list
            // Make sure to prefix the entire search path with A before splitting
            string frag = i == 0 ? "A" + path[0] : path[(i-1)..(i+1)];
            var paths = AllPadPaths[frag];
            if (paths.Count == 1)
                result += PathLength(paths[0], level-1);
            else if (paths.Count == 2) // There's never more than 2
                result += Math.Min(PathLength(paths[0], level-1), PathLength(paths[1], level-1));

        }
        Cache.TryAdd((path, level), result);
        return result;
    }

    /// <summary>
    /// Add shortest paths from each button on the numberpad to every other button
    /// </summary>
    private static void AddNumpadPaths()
    {
        // Could (should?) do this with BFS. But this way seems easy enough.

        foreach (var c0 in new char[11] {'A', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9'})
        {
            foreach (var c1 in new char[11] {'A', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9'})
            {
                if (c0 == c1)
                    continue;

                // Shortest distance is Manhattan distance, but we must avoid the empty space
                // 'A' -> '0' => (2,3) -> (1,3); dx = -1, dy = 0
                // To decide best paths we should avoid zig-zag paths - ie go straight across first then straight up/down, or vice-versa
                List<string> paths = [];
                var (x0, y0) = NumPadPoint(c0);
                var (x1, y1) = NumPadPoint(c1);
                var (dx, dy) = (x1 - x0, y1 - y0);
                // Left/right first, if we can avoid the empty space
                if (dx > 0 /* ok if we're going right */ || !(y0 == 3 && (x0 == 0 || x1 == 0)))
                {
                    var path = "";
                    path += dx < 0 ?  new string('<', -dx) : new string('>', dx);
                    path += dy < 0 ?  new string('^', -dy) : new string('v', dy);
                    path += "A";
                    paths.Add(path);
                }

                // Up/down first, avoiding empty space
                if (dy < 0 /* ok if we're going up */ || !(y1 == 3 && (x0 == 0 || x1 == 0)))
                {
                    var path = "";
                    path += dy < 0 ?  new string('^', -dy) : new string('v', dy);
                    path += dx < 0 ?  new string('<', -dx) : new string('>', dx);
                    path += "A";
                    paths.Add(path);
                }
                AllPadPaths.Add($"{c0}{c1}", paths);
            }
        }
    }

    private static readonly (int, int) InvalidPoint = (-1, -1);
    private static (int X, int Y) NumPadPoint(char c) => c switch {
        'A' => (2,3),
        '0' => (1,3),
        '1' => (0,2),
        '2' => (1,2),
        '3' => (2,2),
        '4' => (0,1),
        '5' => (1,1),
        '6' => (2,1),
        '7' => (0,0),
        '8' => (1,0),
        '9' => (2,0),
        _ => InvalidPoint,
    };

    // Expect P1: 126384 = 68 * 29 + 60 * 980 + 68 * 179 + 64 * 456 + 64 * 379
    // private const string TestInput = """
    // 029A
    // 980A
    // """;
    private const string TestInput = """
    029A
    980A
    179A
    456A
    379A
    """;
}

File: go/days/day21.go
package days

// https://adventofcode.com/2024/day/21 - Keypad Conundrum

import (
	aoc "aoc2024/aocutils"
	"strconv"
	"strings"
)

type d21CacheKey struct {
	path  string
	level int
}

type day21 struct {
	day       int
	buttonMap map[string][]string
	cache     map[d21CacheKey]int64
}

func Day21() {
	d := &day21{21, nil, make(map[d21CacheKey]int64)}
	d.Run()
}

func (d *day21) Run() {
	input, err := aoc.GetDayInputLines(d.day)
	aoc.CheckError(err)

	buttonMap := d.getButtonMap()
	d.buttonMap = buttonMap
	var part1, part2 int64 = 0, 0
	for _, line := range input {
		linenum, _ := strconv.ParseInt(line[:len(line)-1], 10, 64)
		part1 += d.pathLength(line, 3) * linenum
		part2 += d.pathLength(line, 26) * linenum
	}
	aoc.DayHeader(d.day)
	aoc.PrintResult(1, part1)
	aoc.PrintResult(2, part2)
}

func (d *day21) pathLength(path string, level int) int64 {
	if level == 0 {
		return int64(len(path))
	}
	key := d21CacheKey{path, level}
	if v, ok := d.cache[key]; ok {
		return v
	}
	var result int64 = 0
	apath := "A" + path
	for i := range len(path) {
		frag := apath[i : i+2]
		paths := d.buttonMap[frag]
		if len(paths) == 1 {
			result += d.pathLength(paths[0], level-1)
		} else {
			result += min(d.pathLength(paths[0], level-1), d.pathLength(paths[1], level-1))
		}
	}

	d.cache[key] = result
	return result
}

func (d *day21) getButtonMap() map[string][]string {
	result := map[string][]string{
		"AA": {"A"}, "A^": {"<A"}, "A<": {"v<<A"}, "Av": {"<vA", "v<A"}, "A>": {"vA"},
		"^A": {">A"}, "^^": {"A"}, "^<": {"v<A"}, "^v": {"vA"}, "^>": {"v>A", ">vA"},
		"<A": {">>^A"}, "<^": {">^A"}, "<<": {"A"}, "<v": {">A"}, "<>": {">>A"},
		"vA": {"^>A", ">^A"}, "v^": {"^A"}, "v<": {"<A"}, "vv": {"A"}, "v>": {">A"},
		">A": {"^A"}, ">^": {"<^A", "^<A"}, "><": {"<<A"}, ">v": {"<A"}, ">>": {"A"},
	}

	numpad := map[string]aoc.Point2d{
		"7": aoc.NewPoint2d(0, 0), "8": aoc.NewPoint2d(1, 0), "9": aoc.NewPoint2d(2, 0),
		"4": aoc.NewPoint2d(0, 1), "5": aoc.NewPoint2d(1, 1), "6": aoc.NewPoint2d(2, 1),
		"1": aoc.NewPoint2d(0, 2), "2": aoc.NewPoint2d(1, 2), "3": aoc.NewPoint2d(2, 2),
		" ": aoc.NewPoint2d(0, 3), "0": aoc.NewPoint2d(1, 3), "A": aoc.NewPoint2d(2, 3),
	}

	for _, b0 := range []string{"A", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"} {
		for _, b1 := range []string{"A", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"} {
			key := b0 + b1
			if b0 == b1 {
				result[key] = []string{"A"}
				continue
			}
			xy0 := numpad[b0]
			xy1 := numpad[b1]
			dx, dy := xy1.X-xy0.X, xy1.Y-xy0.Y
			result[key] = []string{}

			// Left/right first
			if dx > 0 || !(xy0.Y == 3 && (xy0.X == 0 || xy1.X == 0)) {
				path := ""
				if dx < 0 {
					path += strings.Repeat("<", -dx)
				} else {
					path += strings.Repeat(">", dx)
				}
				if dy < 0 {
					path += strings.Repeat("^", -dy)
				} else {
					path += strings.Repeat("v", dy)
				}
				path += "A"
				result[key] = append(result[key], path)
			}

			// Up/down first
			if !(dx == 0 || dy == 0) { // Only if path is not a single straight line
				if dy < 0 || !(xy1.Y == 3 && (xy0.X == 0 || xy1.X == 0)) {
					path := ""
					if dy < 0 {
						path += strings.Repeat("^", -dy)
					} else {
						path += strings.Repeat("v", dy)
					}
					if dx < 0 {
						path += strings.Repeat("<", -dx)
					} else {
						path += strings.Repeat(">", dx)
					}
					path += "A"
					result[key] = append(result[key], path)
				}
			}
		}
	}
	return result
}
